### PR TITLE
API: Add support for search queries with incremental updates

### PR DESCRIPTION
**Pull request by thielepaul**
Following the discussions in https://github.com/photoprism/photoprism-mobile/issues/78, I created a new backend API for the community-maintained flutter app.
It is a new endpoint so there is no influence on web frontend. 

The concept behind this API is to use the "UpdatedAt" and "DeletedAt" fields in the SQL database provided by Gorm that allow us to fetch all changes since the last request.

A few data updates in the photoprism backend use the UpdateColumn method instead of the Update method and thus, change the data in the database without updating the timestamp. As this would mean that the new API could not get these updates, I replaced all usages of the UpdateColumn method with the Update method.

see https://github.com/photoprism/photoprism-mobile/pull/96 for the implementation of the new API in the app

### CODE REVIEW

**Comment by benmccann** [comment]
I haven't really understood things well enough to know whether this separate API endpoint is a good idea or not. But regarding the feedback on https://github.com/photoprism/photoprism/pull/723 - I think a better way to do paging in that PR might have been to do it by timestamp instead of count. New pictures auto-uploaded from phone will have a recent timestamp. Paging by count that may shift the results, but paging by timestamp should avoid that issue

**Comment by thielepaul** [comment]
> I haven't really understood things well enough to know whether this separate API endpoint is a good idea or not. But regarding the feedback on #723 - I think a better way to do paging in that PR might have been to do it by timestamp instead of count. New pictures auto-uploaded from phone will have a recent timestamp. Paging by count that may shift the results, but paging by timestamp should avoid that issue

The new API gives all database entries sorted such that the oldest comes first and the newest comes last. Changes as such will always be added to the bottom of the list. If I use the `since` parameter I can get all changes that are newer than the newest entry that I have already downloaded.
The count parameter is only there to limit the number of entries in the response and thus its size. The actual "paging", although I am not sure if that is the correct term here, is done based on the timestamp.

**Comment by thielepaul** [comment]
@lastzero did you already have time to look into this PR, do you have feedback for me?

**Comment by lastzero** [comment]
Sorry, no :/ drowning in work... didn't forget about this and all the other PRs!

**Comment by gomme600** [comment]
Any update on this? Have been holding off updating the mobile app but auto uploading is pretty broken on the old version.

**Comment by graciousgrey** [comment]
Unfortunately not, we won't find time to review pull requests before April.

**Comment by gomme600** [comment]
> Unfortunately not, we won't find time to review pull requests before April.

Ok, no problem. Not urgent for me until google photos ends free storage in june, after that I think that the mobile app will become a lot more important to more people!
Thanks for the great work you are all doing here!

**Comment by lastzero** [comment]
Sorry for the wait, I finally have time to look through open pull requests (once again). Probably makes sense to add a third search endpoint in addition to the regular (file based) API and the Geo API for Places. 

Question: If this endpoint only returns updated entries, is sorting then happening in the client code? If true, we may skip sorting on the server and return all matching results at once.

**Comment by thielepaul** [comment]
> Question: If this endpoint only returns updated entries, is sorting then happening in the client code? If true, we may skip sorting on the server and return all matching results at once.
Yes sorting and filtering happens on the client side. We could skip sorting on the server side but then I think we should add an offset parameter to get the remaining results if there are more than `count` results. With the sorting we can just use the last updated instead of offset.

**Comment by benmccann** [comment]
I'm not sure offset alone would work as well. What if a new photo is added in the meantime?

**Comment by thielepaul** [comment]
> I'm not sure offset alone would work as well. What if a new photo is added in the meantime?
Without sorting, an example workflow for the API could be:
- initial query [since=0,count=100,offset=0], server returns first 100 results and current server time (before querying the database), client saves time-initial-query
- follow-up query [since=0,count=100,offset=100] server returns remaining 40 results and current server time
- as 40<100, the client now knows that he has all database entries, at least up to the timestamp that was returned from the server as part of the response to the initial query
- the client wants all updates to the database [since=time-initial-query,count=100,offset=0], the server return all (at most 100) rows that have been updated and current server time, client saves time-last-query
- the client wants all updates to the database [since=time-last-query,count=100,offset=0], the server return all (at most 100) rows that have been updated and current server time

this assumes (both, current implementation and without sorting):
- the server has a steady clock (no jumping backward)
- no rows are deleted from the database (marking them as deleted is fine)

If you can still think of a case where adding a new photo would pose a problem, can you please explain it with an example?

**Comment by spewu** [comment]
> > I'm not sure offset alone would work as well. What if a new photo is added in the meantime?
> 
> Without sorting, an example workflow for the API could be:
> 
> * initial query [since=0,count=100,offset=0], server returns first 100 results and current server time (before querying the database), client saves time-initial-query
> * follow-up query [since=0,count=100,offset=100] server returns remaining 40 results and current server time
> * as 40<100, the client now knows that he has all database entries, at least up to the timestamp that was returned from the server as part of the response to the initial query
> * the client wants all updates to the database [since=time-initial-query,count=100,offset=0], the server return all (at most 100) rows that have been updated and current server time, client saves time-last-query
> * the client wants all updates to the database [since=time-last-query,count=100,offset=0], the server return all (at most 100) rows that have been updated and current server time
> 
> this assumes (both, current implementation and without sorting):
> 
> * the server has a steady clock (no jumping backward)
> * no rows are deleted from the database (marking them as deleted is fine)
> 
> If you can still think of a case where adding a new photo would pose a problem, can you please explain it with an example?

In your example you get 100 results in the first query and 40 results in the second query. Let’s say result A is part of the first 100 results and result B is not. Whithout sorting, which guarantees exist to make sure result A is not getting returned again as part of the 40 results, and result B again not being returned?

If you want all results retuned, you would be relying on an implicit ordering as far as I can see.

**Comment by thielepaul** [comment]
>In your example you get 100 results in the first query and 40 results in the second query. Let’s say result A is part of the first 100 results and result B is not. Whithout sorting, which guarantees exist to make sure result A is not getting returned again as part of the 40 results, and result B again not being returned?
>
>If you want all results retuned, you would be relying on an implicit ordering as far as I can see.

Yes, we would be relying on an implicit ordering, but that is much better from a performance perspective compared to sorting the entries explicitly.

**Comment by thielepaul** [comment]
@lastzero I removed the sorting from endpoint. Anything other remarks or could we merge it like this?

**Comment by graciousgrey** [comment]
We did not forget about this. But at the moment Facial Recognition is consuming all our resources.

**Comment by AliMickey** [comment]
Potential fix yet?

**Comment by v411e** [comment]
@thielepaul I really appreciate your work on photoprism-mobile! :1st_place_medal: 
There are more than 65k pictures in my library and the new API brings huuuuge performance improvements to the app. For the folks waiting on this feature - just try the `thielepaul/photoprism:db-api` docker image as described [here](https://github.com/photoprism/photoprism-mobile). Thank you very much for the time and effort you invested here!

**Comment by pedrommvv** [comment]
Any updates on this ?

**Comment by lastzero** [comment]
https://twitter.com/lastzero/status/1424669417867292673

**Comment by pedrommvv** [comment]
@v411e @thielepaul Here I successfully setup the project locally using the Developer Guide, however I unable to use the `thielepaul/photoprism:db-api` image. 
In the project root, changed the dockerfile, and the docker-compose photoprism-db to use the `thielepaul/photoprism:db-api`, still no success. Still, the mobile app popups API version not compatible. Any hints or should I open a separe Issue Ticket?

**Comment by lastzero** [comment]
Didn't forget about this. Since nobody took over support, bugfixing, multi user support, and facial recognition from us, I still wasn't able to work on the search API — except smaller improvements to find people, and combine terms with &. When implementing the timeline view, we'll probability add an in memory index. Should serve well for syncing native apps too. Give us a bit of time and become a sponsor. That helps the most.

**Comment by pggee7198** [comment]
gzla

**Comment by pggee7198** [comment]
[](https://github.com/photoprism/photoprism/pull/995)

**Review by pggee7198 — APPROVED** [review]

**Comment by tvinhas** [comment]
@lastzero  What kind of sponsored donation would facilitate this PR? I really want to have a native Android app with PhotoPrism to replace my Google Photos/Synology Photos.

**Comment by lastzero** [comment]
developer-salary

**Comment by ohthehugemanatee** [comment]
Thanks for your continued work supporting the branched version, @thielepaul  . Can confirm it works very well for me. There are no conflicts, only code review needed here as far as I can see. 

@lastzero Your comment pushed me over the edge to become a sponsor. Others on the thread: you can support Michael and Theresa with just 2€/mo. Consider what you're saving by using it, or what data you're not giving to the algorithm giants. Definitely worth it for me. :)

**Comment by kvaster** [comment]
Same for me. I'm a sponsor now :)
Also going to try to join development.

**Comment by vmartinv** [comment]
Became sponsor too, the app is amazing and I hope it gets improved in the future!

**Comment by resmh** [comment]
Don't mean to disturb, however I'd really like to thank you all for this remarkable set of applications. I know synchronisation can be quite a hell so special thanks to @thielepaul not only for creating the foundation but also for actually building a mobile application upon it that is capable of performing queries in an independent manner. I've compared numerous photo libraries of both server- and client-centered kind and this is a true killer feature hardly reachable to the latter. Even though I've barely scratched the surface, I'll try to contribute in either way.

**Comment by tobiasfiechter** [comment]
Hi everyone. Thank you for the effort. Merging this PR and support for the app would be really appreciated.

**Comment by iGuan7u** [comment]
This feature could help more developers to enjoy this great app, merging this PR will be appreciative, thank you so much.

**Comment by sorryusernameisalreadytaken** [comment]
I am currently unsure about the status of this pull-request. Is it done or not?
Can somebody with write access merge this? If not, why? @lastzero ?

On the mainpage of Photoprism there is a phone with photoprism on it but there is no pp-app to upload photos automatically...
Bildschirmfoto vom 2022-11-08 21-02-44
Most of the photos worldwide are made with smartphones but photoprism has no app. I think this is one of the last steps to get photoprism become a real competitor to google-photos.

Please merge this pr.

THX for this great project.

**Comment by lastzero** [comment]
So you don't want multi user?

**Comment by sorryusernameisalreadytaken** [comment]
> So you don't want multi user?

I want.... :D Sorry can you explain this a bit more? I really need to understand the problem.

**Comment by lastzero** [comment]
The problem is that I am the only developer and we have hundreds of GitHub Issues and 20,000 users. It depends on those users how many developers we can afford, and right now that's just me, as it has been for the last 5 years. That's the reason why this and many other pull requests haven't been merged yet.

Be aware that we asked if they want us to keep working on their mobile app, they said no, they don't have time and it was just a side project. At a later time, they suggested this API change, but we had already moved on to other topics, like face recognition.

Bottom line:
- If you want me to finish the multi-user feature, you have to be patient, because I can't focus on the multi-user feature, do bugfixes, update the website, write documentation, provide support and extend our API at the same time.
- The combination of PWA and PhotoSync is not bad at all. You might want to give it a try. It's not like you can't use PhotoPrism on a mobile phone. In fact, it works everywhere and not ONLY on iOS or Android. You can even install it on a Windows desktop. How cool is that?
- If you want an app to manage photos on your phone only, then there's a ton of great alternatives. Don't need to reinvent the wheel with apps that already exist.
- If you want us to work faster then join our project or become a sponsor if you have not signed up yet.

**Comment by sorryusernameisalreadytaken** [comment]
THX for the comprehensive reply.

I am sitting here next week with a java developer, an android developer and a dev-ops-enginers. Do you know some "low hanging fruits" we can tackle together to bring this pr forward?

**Comment by lastzero** [comment]
Since our search API is part of the application core and merging it is a commitment to maintain it for years to come, I'm currently the only one who can review it, merge it, and refactor it if needed.

However, you're welcome to take a look at any issues labeled "help wanted", as that will give me time to work on pull requests instead of taking care of small bugs and improvements that could be addressed by contributors.

**Comment by gjedeer** [comment]
I'd like to see this merged as much as everyone else, but until then, I've found this mobile app which runs nicely without any extra APIs: https://radiokot.com.ua/p/photoprism-android-gallery

**Comment by Thalley** [comment]
> I'd like to see this merged as much as everyone else, but until then, I've found this mobile app which runs nicely without any extra APIs: https://radiokot.com.ua/p/photoprism-android-gallery

Cool, thanks! It lacks the one thing I miss, which is the automatic sync support. I have not yet found any free apps that does this well with Photoprism.

**Comment by binaryben** [comment]
> It lacks the one thing I miss, which is the automatic sync support.

Yeh, similar for me. I'm fairly sure the flutter based app referenced in the OP does sync and supports iOS. I was originally waiting for that app to be usable (so by extension, waiting for this issue to be completed). That and multi-user support via OIDC are the two things I needed before implementing a PhotoPrisim instance.

Getting concerned both are going to be behind the outrageously expensive paywall though so might have to start looking for an alternative or consider forking this and stripping the paywalls.

**Comment by ippocratis** [comment]
> > I'd like to see this merged as much as everyone else, but until then, I've found this mobile app which runs nicely without any extra APIs: https://radiokot.com.ua/p/photoprism-android-gallery
> 
> Cool, thanks! It lacks the one thing I miss, which is the automatic sync support. I have not yet found any free apps that does this well with Photoprism. 

Foldersync can do auto sync for free but with adds
Other than that it worth every penny

**Comment by lastzero** [comment]
> That and multi-user support via OIDC are the two things I needed before implementing a PhotoPrisim instance.
>
> Getting concerned both are going to be behind the outrageously expensive paywall though so might have to start looking for an alternative or consider forking this and stripping the paywalls.

How do you think we can release OIDC support and merge this if we can't pay our bills? 🤔

**Comment by binaryben** [comment]
> How do you think we can release OIDC support and merge this if we can't pay our bills? 🤔

I'll take that as admission OIDC is going to be a subscription only feature.

I have no issues with you needing to pay your bills, but your pricing is extreme considering it doesn't include storage and bandwidth. It also seems like development choices are being made that are more focused on profit over quality of product. High detailed maps being paid because it costs you money for example... hmm, seems like an excuse when many users would be ecstatic with OSM based maps.

Time for a community fork potentially. PhotoPrisium anyone?

**Comment by lastzero** [comment]
@binaryben Before you go down that road, you should read this:

- https://www.photoprism.app/oss/faq#funding
- https://www.photoprism.app/code-of-conduct
- https://github.com/photoprism/photoprism/issues/782#issuecomment-1494006370
- https://github.com/photoprism/photoprism/issues/2998

**Comment by binaryben** [comment]
@lastzero 

> Before you go down that road, you should read this:

Trust me, I'd rather not go down that road unless absolutely needed. I am happy to engage any and every conversation that could lead away from that road. I'm creating an opinionated stack for people to be able to self-host and stop needing Google/Apple/Facebook/etc. Much more interested on the orchestration and integration side of hosting multiple services than needing to put effort onto maintaining individual services. Work historically fills my desire for product driven development activities 😅

> * https://www.photoprism.app/oss/faq#funding

Have already read this before my earlier comments. Is there anything in particular you want me to take from it though? As I've said, I have no issue with paying. I pay for several FOSS based services, even when using the self-hosted (usually completely free and unrestricted) version because I believe as you do that projects need funding to be sustainable. But your pricing is very difficult to justify (this comment is directed primarily at the self-hosted options; managed options that need to cover bandwidth, storage, admin and support are a different kettle of fish entirely).

> * https://www.photoprism.app/code-of-conduct

Have also read previously to commenting. Is the reason for linking to it implying I am violating it or encouraging me to contribute to the below issues instead of forking? I am open to contributing but not with the current situation of PRs sitting open for years. A focus on "zero bugs" at the cost of progress is typically a reason to suspect an inexperienced development team and unsustainable underlying business model. I'd be more likely to contribute to Immich where development moves faster in the current state of the two projects as that will ultimately be more likely to lead to a sustainable project/product. Have you considered seeking funding from venture capitalists as an option or crowdsourcing funding?

> * [Auth: OpenID Connect #782 (comment)](https://github.com/photoprism/photoprism/issues/782#issuecomment-1494006370)

The specifically linked comment is the standard boilerplate explanation why things are taking time. Not sure what specifically to take from it. There is no answer anywhere (I have seen at least) on whether OIDC support will be a premium feature or not. Currently users can use the CLI to add users according comments I've read elsewhere so that's good. Convenience features being premium tend to work out fine if you can get similar functionality another way. Typically the users who can use a CLI are also the ones who are more likely to be able to contribute in others ways (i.e. PRs). I'm willing to assume most users who are looking to self host are also using tools like Authelia. I'd be very curious to hear the justification of why OIDC should be a premium feature beyond "because we can and because we need funding".

> * [Places: Improve the level of detail of the included offline world map #2998](https://github.com/photoprism/photoprism/issues/2998)

Yes, this is slightly more interesting in that there is the semblance of a technical driven argument over political. But again, [an easy technical solution was proposed](https://github.com/photoprism/photoprism/issues/2998#issuecomment-1367954835) and [pushed aside by yourself](https://github.com/photoprism/photoprism/issues/2998#issuecomment-1367990366) for political reasons. You're essentially asking your community in the linked issue above to duplicate the work already done by the OSM community. That mentality doesn't tend to sit well with OSS communities. We want to build on what has already been done, not recreate the wheel (yes, I know there are some developers that love to recreate technically superior wheels and we love their work, but there is a time and place for that and this is not it).

Excuse my appearance of ranting. Please take it as evidence of someone who wants your team to succeed and be profitable rather than someone who is entitled. I understand the work that goes into these types of projects.

**Comment by gjedeer** [comment]
> automatic sync support

I just do that with Syncthing. Very reliable.

**Comment by lastzero** [comment]
**Have you considered other funding options like venture capital?**

Yes, we have considered other options many times. Based on [feedback from our community](https://twitter.com/photoprism_app/status/1457281530301145088), we decided to fund development with paid memberships that provide additional features and services. Keep in mind that by accepting venture capital, we would lose our independence, which could ultimately [ruin our mission](https://twitter.com/Abbott/status/1457369715257581568).

:point_right: https://www.photoprism.app/oss/faq#venture-capital

**Comment by binaryben** [comment]
> I just do that with Syncthing. Very reliable.

Yeh, I've also seen a few suggestions of using icloudpd and the google play equivalent options as well. Definitely options although a tad hacky and not the easiest to setup and maintain in household situations (not many want to grab a cookie to get past Apple's 2FA).

**Comment by jurgenhaas** [comment]
> I just do that with Syncthing. Very reliable.

I do that with Nextcloud, no problems at all, either.

**Comment by lastzero** [comment]
> But your pricing is very difficult to justify (this comment is directed primarily at the self-hosted options; managed options that need to cover bandwidth, storage, admin and support are a different kettle of fish entirely).

@binaryben It is a membership fee that allows us to maintain the project for everyone, including users who don't pay anything. If we provide all features for free to self-hosters, we'll have to change our funding model to focus e.g. on a SaaS solution, which goes against our mission to protect your privacy.

That said, we are open to discussing our pricing - just not here.

> Have you considered seeking funding from venture capitalists as an option or crowdsourcing funding?

With venture capital, we lose control and make sure that everything we offer is commercialized. If you have experience in the industry, you should know that.

Our model is basically crowdfunding or sponsorware. That means we plan to release member features to everyone at a later time, which is mentioned in the FAQs and has already happened for some features and config options.

> The specifically linked comment is the standard boilerplate explanation why things are taking time. Not sure what specifically to take from it. There is no answer anywhere (I have seen at least) on whether OIDC support will be a premium feature or not. [...] I'd be very curious to hear the justification of why OIDC should be a premium feature beyond "because we can and because we need funding".

You seem to think that we are making a big profit and OIDC will be developed anyway, which is not the case. Once we have it, we can think about how to make it available.

> Excuse my appearance of ranting. Please take it as evidence of someone who wants your team to succeed and be profitable rather than someone who is entitled. I understand the work that goes into these types of projects.

Giving advice is one thing, but doing so in GitHub Issues (PR comments) in a disrespectful way is something we cannot tolerate. This is because it affects our well-being and keeps us from doing the work you seem to think will get done anyway.

**Comment by binaryben** [comment]
> It is a membership fee that allows us to maintain the project for everyone. Even for those who don't pay anything. If we provide everything for free to self-hosters, we'll have to change our funding model to focus e.g. on a SaaS solution, which goes against our mission to protect your privacy.

I'll respectfully disagree. There are plenty of well known and established privacy focused providers of various services. Hosted solutions does not equate to being anti privacy. Take Protonmail for example. There are numerous ways to implement this. You could take a zero knowledge approach for example and encrypt the photos at rest.

> That said, we are open to discussing our pricing - just not here.

Fair. Please direct me to a better place. I'd prefer somewhere that is still publicly visible for transparency.

> ... If you have experience in the industry, you should know that.

Again, a flawed assumption. Yes there is a risk of that happening. I've even recently [been concerned exactly that was happening in another open source project](https://github.com/directus/directus/discussions/17977#discussioncomment-5567591) currently navigating the same issue of how to be profitable because I have dealt with it. It's not a given though (see same project and linked discussion for more info).

> Our model is basically crowdfunding or sponsorware.

It's an attempt at bootstrapping to be fair. In itself, that's a very noble thing. It's just going to be an uphill battle to convince users to pay $80 in perpetuity to host their own data.

If you want it to be seen as sponsorware or crowdfunded, put bounties on feature requests and let people vote with their wallets. I certainly would for some key features.

> That means we plan to release member features to everyone at a later time, which is mentioned in the FAQs and has already happened for some features and config options.

Neat. That could certainly be communicated more widely and there could be a more defined process around this. It's similar to Directus taking on a BSL which I support.

> You seem to think that we are making a big profit ...

Quite the opposite, I think you're struggling and it's sad to see.

> ... and OIDC will be developed anyway, which is not the case.

I can only infer what will be developed by what is shown in [your roadmap](https://github.com/photoprism/photoprism/projects/5) which clearly shows development of OIDC support is in-progress #782.

> Once we have it, we can think about how to make it available.

I have no idea what to make of this... all that comes to mind is failure to have a plan is a plan to fail. I'd really like this project to succeed and become commercially viable. I am very concerned about the monopoly the likes of Apple and Google are having on everyones most personal and sentimental data.

> Giving advice is one thing, but doing so in GitHub Issues (PR comments) in a disrespectful way is something we cannot tolerate. This is because it affects our well-being and keeps us from doing the work you seem to think will get done anyway.

I'm not being disrespectful, just blunt and to the point. If you need me to soften my communication, I will. Please direct me to where you would like discussion about pricing to take place.

**Comment by lastzero** [comment]
> Quite the opposite, I think you're struggling and it's sad to see. [...] I'm not being disrespectful, just blunt and to the point.

@binaryben It may be that you think this kind of communication is respectful, but it is not and your comments have distracted us from other work. As a result, I have blocked you from commenting until further notice.

The reason why we don't explain our approach more often and in more detail (e.g. in a blog) is simply that our team is too small. However, we know it's important and something we should improve. If we offer everything for free, that won't change though. Of course, we have also considered many other options.

Your comments about pricing suggest that everyone has to sign up for the most expensive tier, which is simply not the case. Those who do sign up give us overwhelmingly positive feedback. If you don't agree with that, you are free to use other software.

Finally, the prices on our website were preliminary. On GitHub, you can currently sign up for $64 (one time). The amount we now quote for PhotoPrism+ when you pay via Stripe is the same, just in Euros. We've also lowered the price for the Essentials tier from EUR 24.90 to EUR 23.00, and are offering a discount to students, teachers, and non-profit organizations.

**Comment by lastzero** [comment]
It would have been appropriate for [binaryben](https://github.com/binaryben) to fully disclose his motivation for creating a fork. On LinkedIn, he describes himself as an entrepreneur and the only obvious reason he can't use our software without OpenID Connect or other member features is that he wants to monetize it as part of his own product:

domostack

**Comment by lastzero** [comment]
@thielepaul My apologies for taking such an unusually long time to provide final feedback! We really appreciate the work that has been put into the mobile app, so we didn't want to just close the PR saying that it can't be implemented this way.

In order to come to a final conclusion, it was important for us to have a meaningful and secure multi-user support ready so we know what is required and could maintain our ability to release incremental updates. It is now clear that we cannot allow clients to perform direct queries that bypass existing filters and security mechanisms.

**I am happy to make suggestions for alternatives if there is a chance that the mobile app will continue to be maintained.** This has not always been the case in the past, even in the early days (as explained above). That's why we moved on to other issues before this PR was submitted. Obviously, the back and forth was confusing for our team as well, but sometimes that's just the way it is.

In any case, this needs to be reworked (e.g. integrated with our existing search API) as security is a top priority for us and many of our users. It therefore seems best to close this PR with an offer to work on an alternative solution if that is an option and there is time and motivation to continue working on this particular mobile app.

To all users who are still following, I would like to say that **this was never in the way of developing a mobile app in general**, as there are many other apps that work flawlessly with PhotoPrism and where the developers never had any problems with the integration, for example [Gallery for PhotoPrism](https://play.google.com/store/apps/details?id=ua.com.radiokot.photoprism&hl=en_US&pli=1) and the apps we feature [on our partners page](https://www.photoprism.app/partners).

In other words, it was completely unnecessary to get angry, and in some cases, extremely disrespectful. That's one more reason to close this PR now and make a fresh start. Thank you very much!
